### ISSUE TITLE
FR: [Request for encrypted PDF support]

### ISSUE THREAD

**Issue by bncbg** [koreader/koreader#10046]
I'm writing to inquire about the support for encrypted PDF files in KOReader.

I have attempted to create sample files in various formats, but I have only been able to successfully open files encrypted with "Acrobat 3.0 and later, 40-bit RC4" encryption:

- [Test-Page (Acrobat 3.0 and later, 40-bit RC4).pdf](https://github.com/bncbg/temp/raw/main/Test-Page%20(Acrobat%203.0%20and%20later%2C%2040-bit%20RC4).pdf)
- [Test-Page (Acrobat 6.0 and later, 128-bit RC4).pdf](https://github.com/bncbg/temp/raw/main/Test-Page%20(Acrobat%206.0%20and%20later%2C%20128-bit%20RC4).pdf)
- [Test-Page (Acrobat 7.0 and later, 128-bit AES).pdf](https://github.com/bncbg/temp/raw/main/Test-Page%20(Acrobat%207.0%20and%20later%2C%20128-bit%20AES).pdf)
- [Test-Page (Acrobat X and later, 256-bit AES).pdf](https://github.com/bncbg/temp/raw/main/Test-Page%20(Acrobat%20X%20and%20later%2C%20256-bit%20AES).pdf)

Password: `test123`

Can you provide information about the current status of support for encrypted PDFs? If KOReader does not currently support opening encrypted PDF files, I request that this feature be developed. As a non-technical end-user, I'm unable to contribute to the development of the language used.

Thank you in advance for your valuable support and I wish the entire team all the best.

**Comment by Frenzie** [comment]
I believe PDF 1.7 Extension Level 8 encryption should be supported as well 256-bit AESv3, which seems like it might match your final file.

The log says there's some issue with that file, but I suppose that's indicative for something that's either newer or otherwise unsupported.

PS I didn't try MuPDF 1.16+ which might yield slightly different results.

```
warning: aes padding out of range
warning: ignoring zlib error: incorrect header check
error: corrupt object stream (292 0 R)
warning: cannot load object (3 0 R) into cache
01/23/23-20:41:09 WARN  cannot open document /home/frans/Downloads/Test-Page (Acrobat X and later, 256-bit AES).pdf frontend/document/pdfdocument.lua:35: ./ffi/mupdf.lua:78: MuPDF found no pages in file.: corrupt object stream (292 0 R) (2)
```

**Comment by NiLuJe** [comment]
Yup, same. Works with the upstream mupdf tools on current releases, FWIW, but, AFAIK, this was implemented early enough that we should support it.

**Comment by NiLuJe** [comment]
> PS I didn't try MuPDF 1.16+ which might yield slightly different results.

It might, the AES stuff is currently inherited from a god-awful minizip version, and I think that was changed.

**Comment by Frenzie** [comment]
Yes, it's been supported since much longer ago but there have been at least some changes and/or bug fixes since.
